### PR TITLE
rc05

### PR DESCRIPTION
**Pull request by heapwolf**
- Improves reflection
- Improves sync
- Improves binary encoding/decoding
- Improves reliability in general
- Introduces queries
- Overall fixes several critical tests in the main suite

### CODE REVIEW

**Review by chicoxyzzy — COMMENTED** [review]
needs linting fixes

**Review by chicoxyzzy — APPROVED** [review]
Other then linting lgtm

**Comment by chicoxyzzy** [comment]
oh actually the description doesn't reflect the changes in files. was the changes merged in another PR? @heapwolf

**Comment by jwerle** [comment]
this needs latest update from the protocol

File: api/stream-relay/index.js [hunk 1]
```diff
@@ -217,9 +235,9 @@ export const wrap = dgram => {
     reflectionFirstResponder = null
     peerId = ''
     isListening = false
-    ctime = Date.now()
+    sdfsactime = Date.now()
```
Comment (chicoxyzzy) [inline api/stream-relay/index.js:238]: is this an abbreviation?

File: api/stream-relay/index.js [hunk 2]
```diff
@@ -231,41 +249,21 @@ export const wrap = dgram => {
...
...11f6caf53d0955db88","natType":31,"indexed":true},{"address":"3.68.89.3","port":22787,"peerId":"448b083bd8a495ce684d5837359ce69d0ff8a5a844efe18583ab000c99d3a0ff","natType":31,"indexed":true},{"address":"3.76.100.161","port":25761,"peerId":"07bffa90d89bf74e06ff7f83938b90acb1a1c5ce718d1f07854c48c6c12cee49","natType":31,"indexed":true},{"address":"3.70.241.230","port":61926,"peerId":"1d7ee8d965794ee286ac425d060bab27698a1de92986dc6f4028300895c6aa5c","natType":31,"indexed":true},{"address":"3.70.160.181","port":41141,"peerId":"707c07171ac9371b2f1de23e78dad15d29b56d47abed5e5a187944ed55fc8483","natType":31,"indexed":true},{"address":"3.122.250.236","port":64236,"peerId":"a830615090d5cdc3698559764e853965a0d27baad0e3757568e6c7362bc6a12a","natType":31,"indexed":true},{"address":"18.130.98.23","port":25111,"peerId":"ba483c1477ab7a99de2d9b60358d9641ff6a6dc6ef4e3d3e1fc069b19ac89da4","natType":31,"indexed":true},{"address":"13.42.10.247","port":2807,"peerId":"032b79de5b4581ee39c6d15b12908171229a8eb1017cf68fd356af6bbbc21892","natType":31,"indexed":true},{"address":"18.229.140.216","port":36056,"peerId":"73d726c04c05fb3a8a5382e7a4d7af41b4e1661aadf9020545f23781fefe3527","natType":31,"indexed":true}]
```
Comment (chicoxyzzy) [inline api/stream-relay/index.js:265]: introducers?

File: api/stream-relay/index.js [hunk 3]
```diff
@@ -432,39 +440,53 @@ export const wrap = dgram => {
...
+        if (expired) { // || !NAT.isValid(peer.natType)) {
+          const p = this.peers.splice(i, 1)
+          if (this.onDisconnect) this.onDisconnect(p)
+          continue
         }
+      }
 
+      // heartbeat
+      const { hash } = await this.cache.summarize('', this.cachePredicate)
+      for (const [, peer] of Object.entries(this.peers)) {
```
Comment (chicoxyzzy) [inline api/stream-relay/index.js:482]: this can be just `Object.values(this.peers)`, but it's also fine

File: api/stream-relay/index.js [hunk 4]
```diff
@@ -488,10 +510,29 @@ export const wrap = dgram => {
      * @ignore
      */
     send (data, port, address, socket = this.socket) {
+      this.sendQueue.push({ data, port, address, socket })
+      this._scheduleSend()
+    }
+
+    _scheduleSend () {
```
Comment (chicoxyzzy) [inline api/stream-relay/index.js:517]: we can add `@private` JSDoc tag here to improve TS definitions a bit, same for `_dequeue`

File: api/stream-relay/index.js [hunk 5]
```diff
@@ -1007,79 +1113,161 @@ export const wrap = dgram => {
...
+      //
+      // receiving an answer
+      //
+      if (this.returnRoutes.has(queryId)) {
+        rinfo = this.returnRoutes.get(queryId)
+
+        let p = packet.copy()
+        if (p.index > -1) p = await this.cache.compose(p)
+
+        if (p && p.index === -1) {
```
Comment (chicoxyzzy) [inline api/stream-relay/index.js:1222]: hmmm is `p &&` necessary here? it will throw on line 1220 if it's not truthy. or maybe it should be in 1220 there as well?

File: api/stream-relay/index.js [hunk 6]
```diff
@@ -1089,30 +1277,22 @@ export const wrap = dgram => {
...
-      }
-
-      if (packet.message.natType) {
-        const peerId = packet.message.requesterPeerId
-        const natType = packet.message.natType
-        peer = this.addPeer({ peerId, natType, port, address })
+      if (isHeartbeat) {
+        // const peer = this.getPeer(packet.message.requesterPeerId)
+        // if (peer && natType) peer.natType = natType
+        return
```
Comment (chicoxyzzy) [inline api/stream-relay/index.js:1292]: seems like this one can be earlier then object destruction on line above

File: api/stream-relay/index.js [hunk 7]
```diff
@@ -1808,26 +2016,22 @@ export const wrap = dgram => {
      */
     async _onMessage (data, { port, address }) {
       const packet = Packet.decode(data)
-      if (!packet || packet.version < VERSION) return
+      if (!packet || packet.version !== VERSION) return
```
Comment (chicoxyzzy) [inline api/stream-relay/index.js:2019]: does it mean that any packet of message that is not compatible with current version will be ignored? this can make it hard for people who don't update their apps frequently have less peers or no peers.

**Review by chicoxyzzy — APPROVED** [review]
It's hard to follow everything in this PR, I've tried my best to look precisely, but mostly found minor things. Overall LGTM

File: api/stream-relay/cache.js [hunk 8]
```diff
@@ -229,14 +286,21 @@ export class Cache {
...
     }))
 
+    let hash
     // compute a summary hash (checksum of all other hashes)
-    const hash = await this.sha1(buckets.join(''), true)
+
+    if (!buckets.every(b => b === null)) {
+      hash = await this.sha1(buckets.join(''), true)
+    } else {
+      hash = 'da39a3ee5e6b4b0d3255bfef95601890afd80709'
```
Comment (bcomnes) [inline api/stream-relay/cache.js:295]: What hash is this?

File: api/stream-relay/index.js (hunk 1 above)
Comment (bcomnes) [reply api/stream-relay/index.js:238]: looks like someone tying asdf haha.

File: api/stream-relay/index.js [hunk 9]
```diff
@@ -231,41 +249,21 @@ export const wrap = dgram => {
...
-    ]`/* snapshot_end=1691579150299 */).map((/** @type {object} */ o) => new RemotePeer({ ...o, indexed: true }, this))
+    gate = new Map()
+    returnRoutes = new Map()
+
+    metrics = {
+      i: { 0: 0, 1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0, 7: 0, 8: 0, REJECTED: 0 },
+      o: { 0: 0, 1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0, 7: 0, 8: 0 }
+    }
+
+    peers = JSON.parse(/* snapshot_start=1691579150299, filter=easy,static */`
```
Comment (bcomnes) [inline api/stream-relay/index.js:264]: How come this is saved as JSON?

File: api/stream-relay/index.js (hunk 5 above)
Comment (bcomnes) [reply api/stream-relay/index.js:1222]: Switch to `p?.index`?

**Review by bcomnes — APPROVED** [review]

**Comment by heapwolf** [comment]
> It's hard to follow everything in this PR, I've tried my best to look precisely, but mostly found minor things. Overall LGTM

no worries, this has already been under review in its repo. @missinglink and I have been going over this for a while.
